### DISCUSSION TITLE
Version 2.0 break changes

### DISCUSSION THREAD

**Discussion by Marmau** [drizzer14/fnts#29] [General]
Hi :)

Just a quick feedback...

I discovered your lib last week in version 1.8. I use your pipe & ternary functions. 
You removed/refactored theses function for type safety reasons... 

But for some cases (the way I use them actually) it's type safe, and I think the new compose/pipe are not very elegant... 

Voila voila, Hope you'll change your mind 😄

**Comment by drizzer14** [comment]
Hi, @Marmau!

Thanks for the feedback, really appreciate it.

Could you share examples of how you've been using `pipe` and `ternary` in your code?

The way I've used them for the last year (almost), I've encountered so much problems that it resulted in 2.0 with these changes you mention here.

I could go in details if you're interested, but this is all I could say for now just to be short.

**Reply by Marmau** [reply]
I have a a bunch of functions 
```
const a = (x: A) => A
const b = (x: A) => A
const c = (x: A) => A
```
I use pipe with my functions and the types stay safe
```
const d = pipe(a, b, c) // Infered (x: A) => A
```

```
const p: (x: A) => boolean
const e = ternary(p, a, b) // Infered (x: A) => A
const f = ternary(p, identity, c) // Infered (x: A) => A
```

But yes... I'm working with only 1 type 🤔

**Reply by drizzer14** [reply]
Okay, thanks, I get it now. Below I'll try to explain what went on with `pipe` and `ternary`, but forgive my simplicity – I'm not that much of an expert here, just playing around mostly.

`pipe` is fine for these simple use cases as you don't have complex types (like unions/intersections), function overloads or generics. Even if you do have those, but your functions are simple and deterministic enough, `pipe` will be able to figure out what inputs and outputs you have all the way through your pipeline (or even through a part of it, if it's enough).
Problems usually arise when having functions in a pipeline which accept a generic type as an input, like `<T>(value: T) => ...`. TypeScript is not able to tell what exactly the type here should be and infers it as `unknown`. With overloads it's even more simple – sometimes (I'm not able to figure out when exactly) TypeScript cannot pick a correct overload for your pipeline which results in it breaking entirely due to this one function being "incorrect".
The further you go into generics/overloads field, the more `pipe` becomes unreliable and fragile.

As for `ternary`, the situation is similar, with two small additions:
1. Almost always you have to specifically annotate input and output types, because our input always comes from returned function, which `ternary` cannot know "in advance", especially with generics and overloads once again.
2. Syntactic ternary operator `? :` understands type guards (like `(value: any): value is string`) and can narrow types down conditionally. `fnts/ternary` cannot do that, and currently we have no ability to make it do so.

Currently I have no intention to put `ternary` back, since in most situations it made my code much more verbose and cluttered, which was not what I wanted from it.
`pipe` probably could make its way back into the library as it was, but under another name (like maybe `flow` or something) and when I figure out what exactly should be done about it to make it work better. Or when TypeScript introduces better inference capabilities for such use cases. Composing pipes by groups of two should be fine for now, I guess.

Hope I made the situation clearer for you, and thanks again for the feedback.

Also, if you have some practical ideas about what I've said here, you're free to submit an issue or a pull request 🙂

**Reply by Marmau** [reply]
Thank you for your clarification!
I assume the best that can happen is an improvement Typescript side...

**Reply by drizzer14** [reply]
Yeah, I guess.
TypeScript is better suited for object-oriented style, rather than functional, so there's not much we can do to make ourselves comfortable with functions here. Nevertheless, I found myself in a situation where I wanted to program in a functional style without much complications and abstractions, like porting an entire Haskell standard library into JavaScript 🙃 Thus, `fnts` was born – out of curiosity and attempts to befriend TypeScript with functions without introducing alien patterns and ways of writing code in a way that is unnatural for the JavaScript world.
